### PR TITLE
[https://nvbugs/6482606][fix] Enforce output length at disaggregated generation first-token handoff

### PR DESCRIPTION
**Pull request by xwang233**
## Dev Engineer Review

- Updated disaggregated generation “context-side first token” handling to apply end-id, output-length (max token / max_seq_len budgeting), and stop-word criteria before scheduling the next generation step, using `TorchSampler` stop-criteria helpers.
- Implemented `_apply_disagg_gen_first_token_stop_criteria` to compute per-beam finish reasons (END_ID first, then max-token/OSL budget, then stop-words) and only transition to `GENERATION_COMPLETE` when **all beams** have finished, preventing premature completion and extra tokens (including the `max_tokens=1` edge case).
- Updated `_handle_first_token_response` to skip emitting/handling a first-token response for disagg-gen requests already marked `is_finished` at injection time (preventing redundant token/final-response double counting) and adjusted `exclude_last_generation_logits` so terminal responses preserve prepended logits.
- Added/expanded unit coverage for the above behavior across disaggregated generation handoff scenarios (including paths with and without speculative decoding, per PR objective), with no public API/config changes.

## QA Engineer Review

Tests added in `tests/unittest/_torch/executor/test_disagg_gen_osl_stop.py`:
- `test_first_token_completes_request_when_max_tokens_is_one`
- `test_first_token_does_not_complete_when_budget_remains`
- `test_first_token_end_id_completes_request`
- `test_first_token_stop_word_completes_request`
- `test_none_max_seq_len_still_enforces_output_length`
- `test_all_beams_complete_together_for_length`
- `test_request_incomplete_until_all_beams_stop`
- `test_handle_first_token_response_skips_finished_request`

Coverage in `tests/integration/test_lists/` (test-db/qa):
- Not found by repo search for `test_disagg_gen_osl_stop.py` or the added test function names.

**Verdict: needs follow-up** — added unit coverage exists, but CI/manual test-list (CBTS) integration coverage status is not confirmed from the available test-list references.

## Description

In disaggregated serving, the context server produces the first output token and hands it to the generation server, which injects it and then schedules a generation step. The injection path in `PyExecutor._prepare_disagg_gen_transmission_complete` appended this first token **without applying any stop criteria**, so a request whose first token already satisfies a stop condition still ran one more generation step and emitted an extra token.

The visible symptom is a request with `max_tokens=1` (OSL=1) receiving **2** output tokens. Aggregated serving does not have this problem because the sampler applies the stop check to the context step's first token and finishes the request before any generation step is scheduled.

The gap only manifests when the first token *alone* satisfies a stop condition (`max_tokens=1`, or the first token is the end id / completes a stop word); for `max_tokens > 1` the per-token stop check already stops generation at exactly the requested length. The affected path is shared by disaggregated generation with and without speculative decoding.

### Fix

- Apply the same stop criteria the aggregated sampler applies (`TorchSampler._handle_stop_criteria`: end id, max-token / length, stop words) to the injected first token, before the generation step runs, and finish the request (per beam; complete only when all beams have stopped) so the sampler, which skips finished requests, does not append an extra token.
- Skip the redundant first-token response for a request already finished at injection, mirroring the existing behavior in `_emit_first_token_responses`; otherwise a finished-at-injection request would emit two final responses and double-count the token.

## Test Coverage

- [x] `tests/unittest/_torch/executor/test_disagg_gen_osl_stop.py` (GPU-free unit tests):
  - `max_tokens=1` finishes at the injected first token (`FinishReason.LENGTH`); `max_tokens=2` does not.
  - First-token end id and single-token stop word each finish the request.
  - `max_seq_len=None` still enforces the per-request output-length budget.
  - Beam width > 1: all beams complete together for length; request stays in progress until every beam has stopped.
  - `_handle_first_token_response` skips a request already finished at injection.

## PR Checklist

- [x] PR description clearly explains what and why.
- [x] PR follows the TRT-LLM coding guidelines.
- [x] Test cases are provided for the new code paths.
- [x] No API changes.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 1]
```diff
@@ -6676,6 +6725,14 @@ def _handle_first_token_response(self, scheduled_batch):
         new_responses = []
         for req in scheduled_batch.generation_requests:
             if req.py_decoding_iter == 1:
+                # A disagg gen request whose context-side first token already met a
+                # stop criterion (e.g. max_tokens=1) is finished at injection time.
+                # Its single, final response is emitted by _handle_responses; emitting
+                # a first-token response here would produce a second final response and
+                # double-count the token (nvbugs/6482606). Mirror the is_finished skip
+                # in _emit_first_token_responses.
+                if req.is_finished:
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:6767]: This continue also skips the `get_latest_logits_unexcluded()` snapshot below, so for a finished-at-injection request with `return_generation_logits=True` under the overlap scheduler the single prepended `first_gen_logits` chunk is never delivered.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 2]
```diff
@@ -5875,8 +5875,57 @@ def _prepare_disagg_gen_transmission_complete(self, scheduled_batch):
...
+
+        Marks the request GENERATION_COMPLETE once all beams have met a stop
+        criterion, so the immediately following generation step does not emit an
+        extra token. See nvbugs/6482606. Reuses the same primitives the aggregated
+        sampler applies in ``TorchSampler._handle_stop_criteria`` (END_ID, then
+        max-token / OSL, then stop-words), in the same order.
+        """
+        # max_seq_len is Optional on PyExecutor; when unset there is no seq-len cap
+        # to apply, only the per-request max_new_tokens budget.
+        max_seq_len = self.max_seq_len if self.max_seq_len is not None else (
```
Comment (Shixiaowei02) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py]: Consider using the engine's net `max_seq_len` here for parity, or add a comment noting the spec-extended cap is intentional.

**Review by Shixiaowei02 — CHANGES_REQUESTED** [review]

**Comment by Shixiaowei02** [comment]
Thanks a lot for this helpful fix! Could you please check whether the comments above make sense? @xwang233

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 2 above)
Comment (xwang233) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py]: Fixed. Now first check `self.model_engine.max_seq_len`, then fallback to `self.max_seq_len`.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py (hunk 1 above)
Comment (xwang233) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:6767]: Fixed by `req.set_exclude_last_generation_logits(False)`

**Comment by xwang233** [comment]
Thanks @Shixiaowei02 for the review! Both comments make sense. I've resolved both issues. Please take another look when you have time.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 3]
```diff
@@ -5895,8 +5895,70 @@ def _prepare_disagg_gen_transmission_complete(self, scheduled_batch):
...
+        # extension, so that value is already the net cap). When both are unset
+        # there is no seq-len cap to apply, only the per-request max_new_tokens
+        # budget.
+        net_max_seq_len = getattr(self.model_engine, "max_seq_len", None)
+        if net_max_seq_len is None:
+            net_max_seq_len = self.max_seq_len
+        max_seq_len = net_max_seq_len if net_max_seq_len is not None else (
+            1 << 62)
+        beam_finish_reasons: List[Optional[FinishReason]] = []
+        for beam in range(0, beam_width):
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:5942]: Could we consolidate the stop-reason evaluation with the aggregated path?

File: tests/unittest/_torch/executor/test_disagg_gen_osl_stop.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    assert req.state == LlmRequestState.GENERATION_COMPLETE
+    req.set_finished_reason.assert_called_once_with(FinishReason.END_ID, 0)
+
+
+def test_first_token_stop_word_completes_request() -> None:
+    # A single-token stop word completed by the first token finishes the request.
+    # This is the parity case that a bare max-token/END_ID check would miss.
+    executor = _make_executor()
+    req = _make_disagg_gen_req(
+        num_tokens=11, orig_prompt_len=10, max_new_tokens=16, stop_words=([99], [1])
```
Comment (cascade812) [inline tests/unittest/_torch/executor/test_disagg_gen_osl_stop.py:120]: It seems TorchSampler._meet_stop_token_criteria requires request.py_stop_words_list to be a list, but test_first_token_stop_word_completes_request passes a tuple

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 5]
```diff
@@ -5895,8 +5895,70 @@ def _prepare_disagg_gen_transmission_complete(self, scheduled_batch):
...
+            if token == req.py_end_id:
+                reason = FinishReason.END_ID
+            elif TorchSampler._meet_max_token_stop_criteria(
+                    req, max_seq_len, beam):
+                reason = FinishReason.LENGTH
+            elif TorchSampler._meet_stop_token_criteria(req, token, beam):
+                reason = FinishReason.STOP_WORDS
+            else:
+                reason = None
+            beam_finish_reasons.append(reason)
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:5953]: When only some injected beams satisfy END_ID or stop-word criteria, the request remains in progress but the stopped beams are not recorded in beam_search_store.first_finish_reasons, which is the mask consumed by beam-search sampling. Those beams can therefore continue decoding. Record each first-token finish reason in the sampler’s per-slot beam state; transition the request to GENERATION_COMPLETE only when all beams are finished.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 6]
```diff
@@ -5895,8 +5895,70 @@ def _prepare_disagg_gen_transmission_complete(self, scheduled_batch):
...
+                # already satisfy a stop criterion (e.g. max_tokens=1, or the token
+                # is the end id / completes a stop word). In aggregated serving the
+                # sampler applies this exact check to the context step's first token
+                # and finishes the request before any generation step is scheduled.
+                # The disagg gen path injects the first token without that check, so a
+                # generation (MTP) step still runs and emits an extra token, and the
+                # request returns OSL+1 tokens (nvbugs/6482606). Apply the same stop
+                # criteria here, before this iteration's forward, so no extra token is
+                # produced (the sampler skips GENERATION_COMPLETE requests).
+                self._apply_disagg_gen_first_token_stop_criteria(
```
Comment (cascade812) [inline tensorrt_llm/_torch/pyexecutor/py_executor.py:5908]: Wondering if it's possible to detect the stop criteria on context workers, so it doesn't have to do KV cache transfer from context worker to generation worker.
